proxy/src/coco/source.rs

```diff
@@ -0,0 +1,452 @@
+use std::collections::hash_map::DefaultHasher;
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::str::FromStr;
+
+use librad::surf;
+use librad::surf::git::git2;
+
+use crate::coco;
+use crate::error;
+
+/// Branch name representation.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Branch(pub(super) String);
+
+impl fmt::Display for Branch {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// Tag name representation.
+///
+/// We still need full tag support.
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct Tag(pub(super) String);
+
+impl fmt::Display for Tag {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+/// Representation of a person (e.g. committer, author, signer) from a repository. Usually
+/// extracted from a signature.
+pub struct Person {
+    /// Name part of the commit signature.
+    pub name: String,
+    /// Email part of the commit signature.
+    pub email: String,
+    /// Reference (url/uri) to a persons avatar image.
+    pub avatar: String,
+}
+
+/// Representation of a code commit.
+pub struct Commit {
+    /// Identifier of the commit in the form of a sha1 hash. Often referred to as oid or object
+    /// id.
+    pub sha1: git2::Oid,
+    /// The author of the commit.
+    pub author: Person,
+    /// The summary of the commit message body.
+    pub summary: String,
+    /// The entire commit message body.
+    pub message: String,
+    /// The committer of the commit.
+    pub committer: Person,
+    /// The recorded time of the committer signature. This is a convenience alias until we
+    /// expose the actual author and commiter signatures.
+    pub committer_time: git2::Time,
+}
+
+impl Commit {
+    /// Returns the commit description text. This is the text after the one-line summary.
+    #[must_use]
+    pub fn description(&self) -> &str {
+        self.message
+            .strip_prefix(&self.summary)
+            .unwrap_or(&self.message)
+            .trim()
+    }
+}
+
+impl From<&surf::git::Commit> for Commit {
```

Comment:
I think I mentioned it in the commit, but for posterity: I think it'd make more sense to own the `surf::Commit` here rather than cloning the fields separately. Worst comes to worse, we just have to clone the reference that uses the `From` call.